### ISSUE TITLE
Track metadata about which users take actions in Pachyderm

### ISSUE THREAD

**Issue by JoeyZwicker** [pachyderm/pachyderm#3069]
**What is the ask?**
I'd like to be able to see which users took certain actions in my system. For example, who created a commit, who updated a pipeline, etc. 

I need ...
Auth currently allows me to control who can do what, but for audit reasons I need to know specifically which user added certain data or made changes to a pipeline. 

**What is the goal / desired outcome?**

Know which actions were taken by users in my system.

**If there is a way to accomplish this today via workaround, what does that require?**

Maybe from the logs, if that has user info included, but that's a fairly hard task.

**Comment by gabrielgrant** [comment]
This sounds strongly related to (if not the same as?) #2468

**Comment by msteffen** [comment]
This looks the same at #2468 to me, and I think in general is a fairly common feature of auth systems

**Comment by jdoliner** [comment]
Adding an initial design for this: I think we can get most of the value of this feature by adding tracking to commits. Not every action in Pachyderm results in a commit, but almost every action does, and I think for the purposes of this issue being able to see who created a commit will be sufficient. The other major thing that users want auditability for is pipeline creation and updates, that works great because since 1.7.0 pipeline specs are actually stored in commits, so we can use this system as the source of truth and then render that information nicely as part of the `PipelineInfo` you get back from `list/inspect-pipeline` 

Which user should be listed as the creator of downstream commits is an interesting question. The following options seem reasonable to me:

1. the user who created the commit that triggered the output
2. the user who created the pipeline that created the output
3. both of the above 2 (creator would become a repeated field)
4. everyone who created a commit in the output commit's provenance, all of the above and then some
5. output commits have no creator

I'm actually thinking that the last one might make the most sense because all of the other options are basically pulling info from some or all of the commit's provenance in as the creator field, which is duplicating data. Instead we should just use provenance to understand where output commits come from, since that points the full picture and lets us have a single source of truth.

Another kind of interesting question is who gets counted as the creator of a commit when the `StartCommit` and `FinishCommit` (and potentially the put-files inbetween) call are done by different users. I think we should just say that the user that calls `StartCommit` is the creator regardless of who makes the other calls. It seems like the sanest option, if we didn't choose it then open commits wouldn't have creators which would be confusing. I think that's probably one of the more important use cases for this feature "We have an open commit that's been sitting here for 3 days, who do I ask to make sure this is safe to delete."

The last thing I'm not 100% sure on (except the things I haven't thought of yet), is what exactly we should store in `CommitInfo`s to reference users, this is probably a question for @msteffen, it looks like in the auth protos we normally just store strings called `username` so that's probably what we'll want to do here? If so that seems simple enough.

**Comment by jdoliner** [comment]
@msteffen good point. One simple way we could achieve that is by adding a secondary index to the commits collection that indexes on user. I think that would allow us to quickly get commits by user. I'm not sure if that would get problematic when we had a very large number of commits for a single user.
